src/rebase.js

```diff
@@ -116,13 +116,22 @@ module.exports = (function(){
   function _fetch(endpoint, options){
     _validateEndpoint(endpoint);
     optionValidators.context(options);
-    optionValidators.then(options);
     options.queries && optionValidators.query(options);
     var ref = firebase.database().ref(endpoint);
     ref = _addQueries(ref, options.queries);
-    ref.once('value', (snapshot) => {
+    return ref.once('value').then(snapshot => {
       var data = options.asArray === true ? _toArray(snapshot) : snapshot.val();
-      options.then.call(options.context, data);
+      if(options.then){
+        options.then.call(options.context, data);
+      }
+      return data;
+    }, err => {
+      //call onFailure callback if it exists otherwise return a rejected promise
+      if(options.onFailure && typeof options.onFailure === 'function'){
+        options.onFailure.call(options.context, err);
+      } else {
```

Comment:
@qwales1 Yeah, now I see what you are trying to avoid. Let's take a look from another perspective. Will other methods (those that simply return whatever Firebase returns, e.g. `ref.set()`, `ref.update()`, etc.) produce the same error if the operation fails? Second, if we use `ref.once('value', onSuccess, onFailure, context)` instead of `ref.once('value').then(...)` here, will it produce such an error on failure? In other words, does underlying Firebase API produce such errors by itself?